<?xml version="1.0"?>
<launch>
  <!-- Arguments -->
  <arg name="initial_robots" default="2"/>
  <arg name="robot_model" default="waffle"/>
  <arg name="font_path" default="$(find robot_swarm_bridge)/config/simple_font.json"/>
  <arg name="arena_width" default="10.0"/>
  <arg name="arena_height" default="10.0"/>
  <arg name="use_gui" default="true"/>
  <arg name="detailed_logging" default="false"/>
  <arg name="robot_ids" default="1" />
  <arg name="backend_url" default="wss://robot.zerav.la/hubs/robot" />
  
  <!-- Load configuration -->
  <rosparam command="load" file="$(find robot_swarm_bridge)/config/params/robot_params.yaml" />
  
  <!-- Launch Gazebo with the arena -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find gazebo_ros)/worlds/empty.world"/>
    <arg name="gui" value="$(arg use_gui)"/>
  </include>
  
  <!-- Load and spawn the arena -->
  <param name="arena_description" 
         command="$(find xacro)/xacro --inorder $(find robot_swarm_bridge)/urdf/turtlebot_arena.xacro"/>
  <node name="spawn_arena" pkg="gazebo_ros" type="spawn_model" 
        args="-urdf -param arena_description -model arena -x 0 -y 0 -z 0" 
        output="screen" />
  
  <!-- Robot Manager node -->
  <node name="robot_manager" pkg="robot_swarm_bridge" type="robot_manager.py" output="screen">
    <param name="robot_model" value="$(arg robot_model)"/>
    <param name="status_update_rate" value="1.0"/>
    <param name="arena_width" value="$(arg arena_width)"/>
    <param name="arena_height" value="$(arg arena_height)"/>
  </node>
  
  <!-- Task Manager node -->
  <node name="task_manager" pkg="robot_swarm_bridge" type="task_manager.py" output="screen">
    <param name="task_check_rate" value="1.0"/>
    <param name="arena_width" value="$(arg arena_width)"/>
    <param name="arena_height" value="$(arg arena_height)"/>
  </node>
  
  <!-- Enhanced Font Drawer node -->
  <node name="font_drawer" pkg="robot_swarm_bridge" type="turtlebot_font_drawer.py" output="screen">
    <param name="robot_namespace_template" value="/robot/{}/"/>
    <param name="font_path" value="$(arg font_path)"/>
    <param name="speed_linear" value="0.2"/>
    <param name="speed_angular" value="0.5"/>
    <param name="position_tolerance" value="0.05"/>
    <param name="angle_tolerance" value="0.1"/>
    <param name="control_rate" value="10"/>
    <param name="arena_width" value="$(arg arena_width)"/>
    <param name="arena_height" value="$(arg arena_height)"/>
    <param name="detailed_logging" value="$(arg detailed_logging)"/>
  </node>
  
  <!-- Spawn initial robots -->
  <node name="initial_robot_spawner" pkg="robot_swarm_bridge" type="initial_robot_spawner.py" output="screen">
    <param name="num_robots" value="$(arg initial_robots)"/>
  </node>
  
  <!-- Launch the bridge node -->
  <node pkg="robot_swarm_bridge" type="bridge.py" name="robot_swarm_bridge" output="screen">
    <param name="robot_ids" value="$(arg robot_ids)" />
    <param name="backend_url" value="$(arg backend_url)" />
    <param name="config_file" value="$(find robot_swarm_bridge)/config/config.yaml" />
  </node>
  
  <!-- RViz for visualization -->
  <node name="rviz" pkg="rviz" type="rviz" 
        args="-d $(find robot_swarm_bridge)/rviz/robot_swarm.rviz"
        if="$(arg use_gui)"/>
</launch>